### PR TITLE
Feature home assistant

### PR DESCRIPTION
**Pull request by micolspitale93**
Adding the package harmoni home assistant to control the smart home devices. This package is a wrapper for the request to Home Assistant open source service: https://www.home-assistant.io/getting-started/

### CODE REVIEW

File: doc/tutorials/Create-New-Package.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+
+
+
+[DEPRECATED]
+In your package you may want to publish data to a topic so that other packages may use it or you may want to subscribe to some data published by other packages. For example you may want the audio stream coming from the *harmoni_microphone* package.
+This approach is deprecated because HARMONI wants to standardize the exchange of messages between services so that it is managed at an higher level.
+
+## README file
+[HIGHLY SUGGESTED]
```
Comment (RMichaelSwan) [inline doc/tutorials/Create-New-Package.md]:
```suggestion
[STRONGLY SUGGESTED]
```

File: doc/tutorials/Create-New-Package.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+| Parameters           | Definition | Values |
+|----------------------|------------|--------|
+|parameter_1           |            |        |
+|parameter_2           |            |        |
+|parameter_3           |            |        |
+```
+
+
+## Tests
+[HIGHLY SUGGESTED]
```
Comment (RMichaelSwan) [inline doc/tutorials/Create-New-Package.md]:
```suggestion
[STRONGLY SUGGESTED]
```

File: doc/tutorials/Create-New-Package.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+There are other services that, once started, keep on running.
+For example, the microphone service, once started, keeps sending audio data.
+These kind of services implement the *start* and *stop* methods.
+
+
+
+
+[DEPRECATED]
+In your package you may want to publish data to a topic so that other packages may use it or you may want to subscribe to some data published by other packages. For example you may want the audio stream coming from the *harmoni_microphone* package.
+This approach is deprecated because HARMONI wants to standardize the exchange of messages between services so that it is managed at an higher level.
```
Comment (RMichaelSwan) [inline doc/tutorials/Create-New-Package.md]: I think the general policy should be: send small commands and info messages through the harmoni service system. If you have high density data (e.g. images and audio), do publishing and subscription as normal, but be sure to follow the harmoni namespace guidelines so that multiple packages can publish/subscribe to known interfaces (especially useful when sibling packages provide the same data, as is the case with different types of STT)

File: doc/tutorials/Create-New-Package.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+The *.test* file is very similar to the *.launch* file that is stored in the *launch* folder.
+*[your_package].test* is a file that connects the configuration parameters with the actual implementation of the service, which is stored in the *src* or *nodes* folders.
+The file *[your_package].test* contains also the link to the actual test file, which is usually called *rostest_[your_package].py*.
+The file *[your_package].test* tries to complete the task written in the *rostest_[your_package].py* file, which can succed or fail.
+
+The structure of the *[your_package].test* file is usually like:
+
+```
+<launch>
+	<rosparam file="$(find harmoni_[your_package])/config/configuration.yaml" subst_value="True"/>
```
Comment (RMichaelSwan) [inline doc/tutorials/Create-New-Package.md]:
```suggestion
    <rosparam file="$(find harmoni_[your_package])/config/configuration.yaml" subst_value="True"/>
```
try not to mix tabs and spaces for indents in files.

File: doc/tutorials/Create-New-Package.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    <param name="instance_id" value="default"/>
+    
+    <node pkg="harmoni_[your_package]" type="[your_package]_service.py" name="harmoni_[your_package]" output="screen"/>
+
+    <test test-name="test_[your_package]" pkg="harmoni_[your_package]" type="rostest_[your_package].py" />
+</launch>
+```
+Parameters can be specified in the *configuration.yaml* file or directly in the *.test* file. For example you could add a line to have an input parameter with the value "Hello".
```
Comment (RMichaelSwan) [inline doc/tutorials/Create-New-Package.md:87]: It would be wise to also note the use of namespaces when including config files. If these are not added, it is possible for concurrently running services to overwrite eachother's params. There's an example in `face_detect_service.launch`

File: harmoni_actuators/harmoni_hass/src/hass_service.py [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+                        rospy.loginfo("Service call failed")
+                        rospy.loginfo(f"Home Assistant's response is {hass_response.text}, with status code {hass_response.status_code}")
+                        rospy.loginfo("Did you put the correct uri and token in the configuration file?")
+                        self.response_received = True
+
+        except rospy.ServiceException:
+            self.start = State.FAILED
+            rospy.loginfo("Service call failed")
+            self.response_received = True
+            self.result_msg = ""
```
Comment (RMichaelSwan) [inline harmoni_actuators/harmoni_hass/src/hass_service.py]: I would include the exception message in the the output and change loginfo to logerror

File: harmoni_actuators/harmoni_hass/test/rostest_hass.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        self.name = ActuatorNameSpace.hass.name + "_" + self.instance_id
+        self.service_client = HarmoniActionClient(self.name)
+        self.client_result = deque()
+        self.service_client.setup_client(self.name, self.result_cb, self.feedback_cb)
+        # NOTE currently no feedback, status, or result is received.
+        rospy.Subscriber(
+            "/harmoni_hass_default/feedback", harmoniFeedback, self.feedback_cb
+        )
+        rospy.Subscriber("/harmoni_hass_default/status", GoalStatus, self.status_cb)
+        rospy.Subscriber("/harmoni_hass_default/result", harmoniResult, self.result_cb)
```
Comment (RMichaelSwan) [inline harmoni_actuators/harmoni_hass/test/rostest_hass.py]: These should not be used since the service client already handles them. In general if you are subscribing to an ActionServer topic directly, you're doing something wrong. Specific note for the goal status: you can simply use `self.service_client.get_state()` any time you want the goal state. An action client status callback has not actually been implemented (feel free to make an update to action_client to add it though!)

For HARMONI, we typically only use the Publish/Subscribe workflow for high density continuous data (images, audio), and in those cases you would be subscribing to some topic not provided by an ActionServer. You can get a little more insight here: https://wiki.ros.org/ROS/Patterns/Communication#Topics_vs_Services_vs_Actionlib...

File: harmoni_actuators/harmoni_tts/aws_tts_service.py [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+import soundfile as sf
+import numpy as np
+import boto3
+import re
+import json
+import ast
+import sys
+
+
+class AWSTtsService(HarmoniServiceManager):
```
Comment (RMichaelSwan) [inline harmoni_actuators/harmoni_tts/aws_tts_service.py:25]: I would change the name to `AWS_TTS_Service` since it's all acronyms.

File: harmoni_actuators/harmoni_tts/aws_tts_service.py [hunk 9]
```diff
@@ -0,0 +1,326 @@
+#!/usr/bin/env python3
```
Comment (RMichaelSwan) [inline harmoni_actuators/harmoni_tts/aws_tts_service.py:1]: Why is this file here? Should be in nodes or src (seems like this is a copy?)

File: harmoni_actuators/harmoni_tts/setup.py [hunk 10]
```diff
@@ -6,7 +6,7 @@
 # fetch values from package.xml
 setup_args = generate_distutils_setup(
     packages=['harmoni_tts'],
-    package_dir={'': 'src'},
+    package_dir={'': 'nodes'},
```
Comment (RMichaelSwan) [inline harmoni_actuators/harmoni_tts/setup.py]: You can just remove these lines unless you plan to make a lib version that gets imported into the node. Typically nodes are not included in setup as we do not import nodes into other nodes, and therefore they do not need to be installed like libraries do.

File: harmoni_core/harmoni_pattern/nodes/sequential_pattern.py [hunk 11]
```diff
@@ -142,6 +143,12 @@ def _feedback_callback(self, feedback):
     def _detecting_callback(self, data, service_name):
         """Store data from detection to client_results dictionary"""
         data = data.data
+
+        # rospy.loginfo("--------")
+        # rospy.loginfo("Data from detector " + data)
+        # rospy.loginfo("time-> " + str(time()))
+        # rospy.loginfo("--------")
+
```
Comment (RMichaelSwan) [inline harmoni_core/harmoni_pattern/nodes/sequential_pattern.py]: Maybe change things like these to logdebug instead of commenting out?

File: harmoni_core/harmoni_pattern/setup.py [hunk 12]
```diff
@@ -6,8 +6,8 @@
 # fetch values from package.xml
 setup_args = generate_distutils_setup(
     # scripts=[''],
-    # packages=['harmoni_pattern'],
-    # package_dir={'': 'src'},
+    packages=['harmoni_pattern'],
+    package_dir={'': 'nodes'},
```
Comment (RMichaelSwan) [inline harmoni_core/harmoni_pattern/setup.py]: Again: Nodes aren't installed, remove this.

File: harmoni_detectors/harmoni_stt/test/test_google_stt.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+class TestGoogleStt_Valid(TestGoogleStt_Common):
+    def test_IO(self):
+        rospy.loginfo(
+            "TestGoogleStt[TEST]: basic IO test to ensure data "
+            + "('hello' audio) is received and responded to. Waiting for transcription..."
+        )
+        while not rospy.is_shutdown() and not self.result:
+            self.rate.sleep()
+        assert self.result == True
+
```
Comment (RMichaelSwan) [inline harmoni_detectors/harmoni_stt/test/test_google_stt.py:117]: I would add another test to verify the text received is the same as what is expected. All you need is another function with the `test_` prefix. Note that the context will be reset between tests and setup will run again.

**Review by RMichaelSwan — CHANGES_REQUESTED** [review]
Overall good. There are a few minor issues to fix. Also recommend merging in current develop branch changes into this branch before we do the final merge.

**Comment by RMichaelSwan** [comment]
The fixes are good. Unresolved comments still need to be handled or addressed (feel free to disagree with me) and merge conflicts fixed (can be done by merging in develop from the main repo-- @micolspitale93 you might want to do this to your fork's develop first and then have changes from that develop merged to the current branch; up to you, feel free to ping me if you need assistance). Feel free to start discussion or message affected parties (can use git blame to see who last made changes to affected files/lines) if there's something difficult to handle in the conflicts.

File: harmoni_actuators/harmoni_tts/aws_tts_service.py (hunk 9 above)
Comment (etoscano) [reply harmoni_actuators/harmoni_tts/aws_tts_service.py:1]: I have no clue about why this is here. I think this is a mistake. This file can be removed.

File: harmoni_detectors/harmoni_stt/test/test_google_stt.py (hunk 13 above)
Comment (etoscano) [reply harmoni_detectors/harmoni_stt/test/test_google_stt.py:117]: Thank you for your suggestion! I was a bit stuck. 
However, I have just found out that the code the code here for the stt service is not the same as the one in feature/edison-smarthome. I changed it a while ago because it was giving me "ERROR: Broken Pipe" when using the method transcribe_stream_request. So I followed the instruction on the google stt page to create a different method for a continuous stream from the mic.

TLDR: I cannot add another test since this stt service is not working (for me at least).
I can add a test if I change the stt service file to the one in feature/edison-smarthome, where there is a different method for handling a stream.
https://github.com/micolspitale93/HARMONI/blob/feature/edison-smarthome/harmoni_detectors/harmoni_stt/nodes/google_service.py

File: harmoni_actuators/harmoni_tts/aws_tts_service.py (hunk 8 above)
Comment (etoscano) [reply harmoni_actuators/harmoni_tts/aws_tts_service.py:25]: Thank you for the suggestion, I changed the name for the file in /harmoni_tts/nodes

File: harmoni_actuators/harmoni_hass/test/rostest_hass.py (hunk 7 above)
Comment (etoscano) [reply harmoni_actuators/harmoni_hass/test/rostest_hass.py]: Thank you for the correction! I have removed the subscription part. 
Feel free to tell me if I have to remove some other lines.

**Review by jayy7312 — APPROVED** [review]
